File: Coding_231201 Graph/최원준/이강훈/5번.java
/*
https://leetcode.com/problems/find-eventual-safe-states/

아이디어
out 방향 없는 노드부터 BFS 시작
 -> 터미널 노드로 들어오는 노드들의 out 방향 목록에서 해당 터미널 노드를 제거 ( BFS 반복 )
  -> out 방향 목록 비워지는 노드들만 같은 작업 반복

자료구조
큐, HashMap

시간복잡도
O(N)
*/

import java.util.*;

public class Solution {
    public List<Integer> eventualSafeNodes(int[][] graph) {
        Map<Integer, List<Integer>> outMap = new HashMap<>();
        Map<Integer, List<Integer>> inMap = new HashMap<>();

        for (int i = 0; i < graph.length; i++) {
            outMap.put(i, new ArrayList<>());
            inMap.putIfAbsent(i, new ArrayList<>());

            for (int adj : graph[i]) {
                outMap.computeIfAbsent(i, key -> new ArrayList<>()).add(adj);
                inMap.computeIfAbsent(adj, key -> new ArrayList<>()).add(i);
            }
        }

        Queue<Integer> queue = new LinkedList<>();

        for (Map.Entry<Integer, List<Integer>> entry : outMap.entrySet()) {
            if (entry.getValue().size() == 0) queue.add(entry.getKey());
        }

        while (!queue.isEmpty()) {
            int node = queue.poll();

            if (outMap.get(node).size() == 0) {
                for (int inNode : inMap.get(node)) {
                    outMap.get(inNode).remove((Integer) node);

                    if (outMap.get(inNode).size() == 0) queue.add(inNode);
                }
            }
        }

        List<Integer> answer = new ArrayList<>();
        for (Map.Entry<Integer, List<Integer>> entry : outMap.entrySet()) {
            if (entry.getValue().size() == 0) answer.add(entry.getKey());
        }

        return answer;
    }
}

File: Coding_231201 Graph/최원준/최원준/5번.py
# https://leetcode.com/problems/find-eventual-safe-states/description/

'''
1. 아이디어 :
    dfs를 사용한다.
    safe node를 찾아야하는데 n번쨰 노드부터 시작해서 dfs를 돌린다.
    dfs메서드에서는 무한 루프를 방지하기 위해 조건문을 걸어놓고,
    safe하지 않는다고 가정을 한다음에 시작을 한다.
    노드의 이웃들이 safe하다면 해당 노드도 safe.
2. 시간복잡도 :
    O(edges + nodes)
3. 자료구조 :
    해시맵
'''

class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        n = len(graph)
        safe = {}

        def dfs(i):
            if i in safe:
                return safe[i]
            safe[i] = False
            for neighbor in graph[i]:
                if not dfs(neighbor):
                    return False
            safe[i] = True
            return True

        res = []
        for i in range(n):
            if dfs(i):
                res.append(i)
        return res
